crates/holochain_sqlite/src/table.rs

```diff
@@ -0,0 +1,322 @@
+    ValidationReceipts,
+    /// Single store for all known agents on the network
+    Agent,
+
+    #[cfg(feature = "test_utils")]
+    TestSingle(String),
+
+    #[cfg(feature = "test_utils")]
+    TestMulti(String),
+}
+
+impl TableName {
+    /// Associates a [TableKind] to each [TableName]
+    pub fn kind(&self) -> TableKind {
+        use TableKind::*;
+        use TableName::*;
+        match self {
+            ElementVaultPublicEntries => Single,
+            ElementVaultPrivateEntries => Single,
+            ElementVaultHeaders => Single,
+            MetaVaultSys => Multi,
+            MetaVaultLinks => Single,
+            MetaVaultMisc => Single,
+            ChainSequence => Single, // int
+            ElementCacheEntries => Single,
+            ElementCacheHeaders => Single,
+            MetaCacheSys => Multi,
+            MetaCacheLinks => Single,
+            MetaCacheStatus => Single,
+            ConductorState => Single,
+            Wasm => Single,
+            DnaDef => Single,
+            EntryDef => Single,
+            AuthoredDhtOps => Single,
+            IntegratedDhtOps => Single,
+            IntegrationLimbo => Single,
+            ValidationLimbo => Single,
+            ValidationReceipts => Multi,
+            Agent => Single,
+            #[cfg(feature = "test_utils")]
+            TestSingle(_) => Single,
+            #[cfg(feature = "test_utils")]
+            TestMulti(_) => Multi,
+        }
+    }
+}
+
+impl ToSql for TableName {
+    fn to_sql(&self) -> Result<rusqlite::types::ToSqlOutput<'_>> {
+        Ok(rusqlite::types::ToSqlOutput::Owned(
+            format!("{}", self).into(),
+        ))
+    }
+}
+
+pub enum TableKind {
+    Single,
+    Multi,
+}
+
+impl TableKind {
+    pub fn is_single(&self) -> bool {
+        matches!(self, Self::Single)
+    }
+
+    pub fn is_multi(&self) -> bool {
+        matches!(self, Self::Multi)
+    }
+}
+
+pub(crate) fn initialize_table_single(
+    conn: &mut Connection,
+    table_name: String,
+) -> rusqlite::Result<()> {
+    // create table
+    conn.execute(
+        &format!(
+            "CREATE TABLE IF NOT EXISTS {} (
+                key       BLOB PRIMARY KEY,
+                val       BLOB NOT NULL
+            );",
+            table_name
+        ),
+        NO_PARAMS,
+    )?;
+
+    // create index
+    let index_name = format!("{}_idx", table_name);
+    conn.execute(
+        &format!(
+            "CREATE INDEX IF NOT EXISTS {} ON {} ( key );",
+            index_name, table_name
+        ),
+        NO_PARAMS,
+    )?;
+    Ok(())
+}
+
+pub(crate) fn initialize_table_multi(
+    conn: &mut Connection,
+    table_name: String,
+) -> rusqlite::Result<()> {
+    // create table
+    conn.execute(
+        &format!(
+            "CREATE TABLE IF NOT EXISTS {} (
+                key       BLOB NOT NULL,
+                val       BLOB NOT NULL
+            );",
+            table_name
+        ),
```

Comment:
```suggestion
        &format!(
            "CREATE TABLE IF NOT EXISTS {} (
                key       BLOB NOT NULL,
                val       BLOB NOT NULL,
                CONSTRAINT {}_pk PRIMARY KEY (key, val)
            );",
            table_name,
            table_name,
        ),
```

Then delete the composite index below, since the compound primary key will be implicitly indexed.

Reply:
Nice, thanks